### PR TITLE
Improve WTextField

### LINKED ISSUE #139: WTextFields don't have the full functionality of a vanilla text field
Unlike the vanilla anvil or command block text fields, WTextFields don't scroll horizontally when it is full of characters. It still allows characters to be typed, but they cannot be seen. Also the delete key has the same functionality as the backspace when used in a WTextField.

Unless there is an option I am unaware of, there is no way to make WTextFields scroll horizontally. It should be noted that the WTextFields I've tested on are inside of a WScrollPanel, so maybe that has something to do with it? The delete key works as normal when used inside a vanilla anvil or command block, it only has the backspace functionality in a WTextField, instead of deleting the character after the cursor.

### PR DESCRIPTION
**Pull request by romangraef**
Potentially fixes #139

Functionality changes made to `WTextField`:

 - Introduce text scrolling
 - Fix CTRL+Movement key movement
 - General refactoring

### CODE REVIEW

**Comment by romangraef** [comment]
This shouldn't have any breaking API changes, but at this point im a bit too tired to be sure :sweat_smile:

File: src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java [hunk 1]
```diff
@@ -33,70 +33,90 @@
 import java.util.function.Predicate;
 
 public class WTextField extends WWidget {
-	public static final int OFFSET_X_TEXT = 4;
```
Comment (Juuxel) [inline src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:36]: Removing public fields/methods is technically api breaking, so could you readd this with `@Deprecated(forRemoval = true)`?

File: src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java [hunk 2]
```diff
@@ -106,112 +126,104 @@ public int getMaxLength() {
...
 		return text.substring(start, end);
 	}
-	
+
 	public boolean isEditable() {
 		return this.editable;
 	}
 
 	@Environment(EnvType.CLIENT)
-	protected void renderTextField(MatrixStack matrices, int x, int y) {
```
Comment (Juuxel) [inline src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:131]: Same here

File: src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java [hunk 3]
```diff
@@ -446,41 +495,29 @@ public void addNarrations(NarrationMessageBuilder builder) {
...
+	 * From an X offset past the left edge of a TextRenderer.draw, finds out what the closest caret position (division
+	 * between letters) is.
+	 *
 	 * @param s
 	 * @param x
+	 *
 	 * @return
 	 */
 	@Environment(EnvType.CLIENT)
+	@Deprecated
```
Comment (Juuxel) [inline src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:507]:
```suggestion
	@Deprecated(forRemoval = true)
```

File: src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java [hunk 4]
```diff
@@ -446,41 +495,29 @@ public void addNarrations(NarrationMessageBuilder builder) {
...
 	}
-	
-	/**
-	 * From a caret position, finds out what the x-offset to draw the caret is.
-	 * @param s
-	 * @param pos
-	 * @return
-	 */
-	@Environment(EnvType.CLIENT)
-	public static int getCaretOffset(String s, int pos) {
```
Comment (Juuxel) [inline src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:479]: Also here

**Review by Juuxel — CHANGES_REQUESTED** [review]
Thanks for this PR! I just have some minor comments. Also, it'd be great if you could revert the formatting changes since the diff is hard to read currently because of them.

File: src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java (hunk 2 above)
Comment (romangraef) [reply src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:131]: renderTextField still exists and just does more stuff now (e.g. render cursor blinking and selection and all that stuff i added.)

Comment (Juuxel) [reply src/main/java/io/github/cottonmc/cotton/gui/widget/WTextField.java:131]: Ah sorry, I didn't notice

**Review by Juuxel — APPROVED** [review]

**Comment by Juuxel** [comment]
Thanks for this, I'm going to clean up some rough edges here and merge it to the 1.18 pres too a bit later.

**Pull request merged**